dali/tensorflow/dali_dataset.cc

```diff
@@ -0,0 +1,174 @@
+  }
+
+  void MakeDataset(OpKernelContext* context, DatasetBase** output) override {
+    *output =
+        new Dataset(context, &serialized_pipeline_, &shapes_, &dtypes_);
+  }
+
+ private:
+  class Dataset : public DatasetBase {
+   public:
+    explicit Dataset(OpKernelContext* ctx, const std::string* serialized_pipeline,
+        const std::vector<TensorShape>* shapes, const DataTypeVector* types)
+        : DatasetBase(DatasetContext(ctx)),
+          dtypes_{types},
+          shapes_{std::vector<PartialTensorShape>(shapes->size(), PartialTensorShape())} {
+    }
+
+    std::unique_ptr<IteratorBase> MakeIteratorInternal(
+        const string& prefix) const override {
+      return std::unique_ptr<IteratorBase>(
+          new Iterator({this, strings::StrCat(prefix, "::DALIDataset")}));
+    }
+
+    const DataTypeVector& output_dtypes() const override {
+      return *dtypes_;
+    }
+
+    const std::vector<PartialTensorShape>& output_shapes() const override {
+      return shapes_;
+    }
+
+    string DebugString() const override { return "DALIDatasetOp::Dataset"; }
+
+   protected:
+    Status AsGraphDefInternal(SerializationContext* ctx,
+                              DatasetGraphDefBuilder* b,
+                              Node** output) const override {
+      return Status::OK();
+    }
+
+   private:
+    class Iterator : public DatasetIterator<Dataset> {
+     public:
+      explicit Iterator(const Params& params)
+          : DatasetIterator<Dataset>(params) {}
+
+      Status GetNextInternal(IteratorContext* ctx,
+                             std::vector<Tensor>* out_tensors,
+                             bool* end_of_sequence) override {
+        return Status::OK();
+      }
+
+     protected:
+      Status SaveInternal(IteratorStateWriter* writer) override {
+        return Status::OK();
+      }
+
+      Status RestoreInternal(IteratorContext* ctx,
+                             IteratorStateReader* reader) override {
+        return Status::OK();
+      }
+
+     private:
+      Status SetupStreamsLocked(Env* env) EXCLUSIVE_LOCKS_REQUIRED(mu_) {
+        return Status::OK();
+      }
+
+      void ResetStreamsLocked() EXCLUSIVE_LOCKS_REQUIRED(mu_) {
+      }
+
+      mutex mu_;
+    };
+
+    const DataTypeVector* dtypes_;
+    const std::vector<PartialTensorShape> shapes_;
+  };
+
+  std::string serialized_pipeline_;
+  std::vector<TensorShape> shapes_;
+  DataTypeVector dtypes_;
+};
+
+REGISTER_OP("DALIDataset")
+    .Attr("serialized_pipeline: string")
+    .Attr("shapes: list(shape) >= 1")
+    .Attr("dtypes: list({float, int32, int64, half}) >= 1")
+    .Output("handle: variant")
+    .SetIsStateful()  // TODO: Is it necessary in order to prevent folding?
+    .SetShapeFn([](tensorflow::shape_inference::InferenceContext *c) {
+      std::vector<tensorflow::PartialTensorShape> shapes;
+      TF_RETURN_IF_ERROR(c->GetAttr("shapes", &shapes));
+      for (unsigned int i = 0; i < shapes.size(); ++i) {
```

Comment:
Maybe we can reuse this function from the present daliop?

Reply:
I think we should keep as much code common as possible, because no matter if dataset or custop op we want to have similar or the same API.

Reply:
I had to put it back, as moving it to a header introduced build problems. Ideally it should be extracted